### PR TITLE
added CLR callback, test

### PR DESCRIPTION
**Pull request by bckenstler**
Callback implementation of a cyclical learning rate policy, as described in Leslie Smith's paper [Cyclical Learning Rates for Training Neural Networks](https://arxiv.org/pdf/1506.01186.pdf).

### CODE REVIEW

**Comment by ahundt** [comment]
@bckenstler can you add an example such as a cifar example? Ideally it would clearly illustrate the benefits this policy provides vs one of the existing cifar examples.

**Comment by bckenstler** [comment]
@ahundt The [paper](https://arxiv.org/pdf/1506.01186.pdf) shows, among other things, that convergence is faster using CLR as opposed to static / exponential learning rates. 

From paper:
screen shot 2017-10-15 at 10 47 11 am

Are you requesting that I reproduce their results?

**Comment by ahundt** [comment]
Yes, you'd just need to modify an existing example like [cifar_resnet.py](https://github.com/fchollet/keras/blob/master/examples/cifar10_resnet.py) or [the examples in keras-contrib](https://github.com/farizrahman4u/keras-contrib/tree/master/examples) that reproduces the result. This would make it easier for users to try it out and incorporate into their own code.

**Comment by ps48** [comment]
Hi @bckenstler @ahundt, I was just going through the PR this looks great to me. CLR is useful and looks generic enough to be applied across Datasets and Nets. Is it just an example of CIFAR-10 that is holding this? If so I can surely help

**Comment by bckenstler** [comment]
@ps48 you are correct, I'd welcome cooperation on this!

**Comment by ps48** [comment]
@bckenstler Awesome! let me have a look in detail 👍

**Comment by ps48** [comment]
@ahundt I'll start implementing the example today. Will also look into changes reviewed!

**Comment by Axel13fr** [comment]
@ahundt you may want to have a look at this video, I believe Jeremy Howard demonstrates it well (and almost shows off with his own fastai lib) http://course.fast.ai/lessons/lesson2.html

**Comment by ahundt** [comment]
@Axel13fr Thanks, but an example is for users, not me. :-)

Demos provide the following services to users:
1. Something easy to run that shows at least one use case that demonstrates it works correctly and how to use it.
2. Users can copy/paste it into their own code can easily adapt it to their use case.

**Comment by titu1994** [comment]
@bckenstler Sorry for the delay in reviewing this PR. A year is simply too long.

If possible, could you make the changes that have been discussed by @ahundt, and adapt one if the cifar or mnist examples from this repo to illustrate basic usage of the call back ?

File: keras_contrib/callbacks/cyclical_learning_rate.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    per-cycle basis.
+    This class has three built-in policies, as put forth in the paper.
+    "triangular":
+        A basic triangular cycle w/ no amplitude scaling.
+    "triangular2":
+        A basic triangular cycle that scales initial amplitude by half each cycle.
+    "exp_range":
+        A cycle that scales initial amplitude by gamma**(cycle iterations) at each
+        cycle iteration.
+    For more detail, please see paper.
```
[the remark this answers is not in the document]
Comment (bckenstler) [reply keras_contrib/callbacks/cyclical_learning_rate.py]: Added link.

File: tests/keras_contrib/callbacks/cyclical_learning_rate_test.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+        def scale_fn(i):
+            return 0.001 + (0.006 - 0.001) * \
+                np.maximum(0, (1 - x(i))) * 1 / (5**(i * 0.0001))
+        for i in range(8000):
+            custom_range.append(scale_fn(i + 1))
+
+        assert_allclose(clr.history['lr'], np.array(custom_range))
+
+    X = np.random.rand(4000, 10)
```
[the remark this answers is not in the document]
Comment (bckenstler) [reply tests/keras_contrib/callbacks/cyclical_learning_rate_test.py]: This looks fine!

File: keras_contrib/callbacks/cyclical_learning_rate.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+            lower boundary in the cycle.
+        max_lr: upper boundary in the cycle. Functionally,
+            it defines the cycle amplitude (max_lr - base_lr).
+            The lr at any cycle is the sum of base_lr
+            and some scaling of the amplitude; therefore
+            max_lr may not actually be reached depending on
+            scaling function.
+        step_size: number of training iterations per
+            half cycle. Authors suggest setting step_size
+            2-8 x training iterations in epoch.
```
[the remark this answers is not in the document]
Comment (bckenstler) [reply keras_contrib/callbacks/cyclical_learning_rate.py]: I included sample parameters in the example in the docstring for CIFAR-10 as mentioned in the paper.

**Comment by bckenstler** [comment]
@ahundt I've made the changes you've requested and included an example for CIFAR-10 in comparison to an existing Keras CIFAR-10 [example](https://github.com/keras-team/keras/blob/master/examples/cifar10_cnn.py). 

@titu1994 In regards to the paper, this callback allows for custom cyclic policies to be made based on either epoch or iteration, so this should be easily adaptable.

**Comment by titu1994** [comment]
@bckenstler Seems PEP8 failed in the new example you added, could you check that? Also, great job on this. From my side, no more changes are required.

@ahundt If you could give a final review, that would be great.

**Comment by bckenstler** [comment]
@titu1994 Formatted! Glad to get this submitted!

**Comment by ahundt** [comment]
Cool, LGTM. Thanks for the contribution!
